### DISCUSSION TITLE
[@types/hast] Element data too restrictive

### DISCUSSION THREAD

**Discussion by Caellian** [DefinitelyTyped/DefinitelyTyped#71296] [Issues with a @types package] [answered]
## Context
`Element` has `data` of type:

```ts
export interface Element extends Parent {
    // ...
    data?: ElementData | undefined;
}
```

Which forces following  `ElementData`:

```ts
interface ElementData {
    // ...
    meta?: string | null | undefined
}
```

## Problem

This prevents `{data: { thisShould: "be valid" }, ...}`.

[Spec](https://github.com/syntax-tree/unist?tab=readme-ov-file#data) says:
> Data represents information associated by the ecosystem with the node.
> 
> This space is guaranteed to never be specified by unist or specifications implementing unist.

But it _is_ specified, and too restrictive.

## Instead it should be
```ts
export interface Element extends Parent {
    // ...
    data?: (ElementData & Record<string, unknown>) | undefined;
}
```

[a comment by an automated account is not shown]

**Comment by ChristianMurphy** [comment, marked as answer]
The design is intentional to allow strong typing.
You can extend it with new attributes.
